### DISCUSSION TITLE
Directory for #include preprocessor

### DISCUSSION THREAD

**Discussion by Shimushushushu** [fortran-lang/fortls#343] [Q&A]
I'm working on a framework ([Flexi](https://github.com/flexi-framework/flexi)) which uses lots of preprocessors, e.g., `#include`.  The framework has a structure like:

```
src/flexi.h
src/a_folder/whatever.f90
```

where `whatever.f90` includes `flexi.h` by `#include "flexi.h"`.  While `gfortran` works if `-Isrc` is present, it seems that fortls cannot find the header.  If I change it to `#include "../flexi.h"`, it does work, but there are so many files that I think I should not do that.

I tried both `source_dirs` and `include_dirs`, by adding `src/**`, but neither works.

Have you ever met the problem, and how did you get over it?

**Comment by gnikit** [comment]
My honest opinion is that a tool should never dictate how your code looks like, so you definitely shouldn't edit your source code to get fortls to work.

From what I understand your settings should look something like this

```json
{
  "include_dirs": "src",
  "pp_suffixes": [".h"],
  "incl_suffixes": [".h"]
}
```

I don't have access to a computer right now, so I'm not sure about `incl_suffixes`.
Let me know how that works.

**Reply by Shimushushushu** [reply]
It does not work.  I don't know whether the output from VS Code helps:

```
["INFO" - 2:07:40 AM] Restarting language server...
["INFO" - 2:07:40 AM] Initialising Language Server for workspace: /path/to/project/test.f90 with command-line options: --enable_code_actions, --hover_signature, --use_signature_help, --lowercase_intrinsics, -c, ./fortls.json, --nthreads=4, --incremental_sync
[INFO - 02:07:41] fortls - Fortran Language Server 2.13.0 Initialized
```

where `./fortls.json` contains the configuration you mentioned above.

**Reply by Shimushushushu** [reply]
I tried to examine the code, and my conclusion is: `parse_fortran.py` works as expected; it recognizes `include_dirs`.  However, whenever I hover on the wavy line below the `#include`, a `textDocument/hover` will send to fortls, which returns

```
RECV {
  "jsonrpc":"2.0",
  "id":7,
  "method":"textDocument/codeAction",
  "params":{
    "textDocument":{
      "uri":"file:///path/to/project/test.f90"
    },
    "range":{
      "start":{
        "line":1,
        "character":2
      },
      "end":{
        "line":1,
        "character":2
      }
    },
    "context":{
      "diagnostics":[
        {
          "range":{
            "start":{
              "line":1,
              "character":2
            },
            "end":{
              "line":1,
              "character":2
            }
          },
          "message":"flexi.h: No such file or directory",
          "severity":1
        }
      ],
      "only":[
        "quickfix"
      ],
      "triggerKind":1
    }
  }
}
```

Why `serve_hover` returns a `codeAction`?  I know nothing of LSP and I'm really confused.

By the way, if I use `"include_dirs": "src"`, it becomes `['s', 'r', 'c']`, so I changed it to `"include_dirs": ["src"]`.

**Reply by zbinkz** [reply]
Does using the full path to `src` in `include_dirs` work?
`"include_dirs": ["/full/path/to/src"]`

**Reply by Shimushushushu** [reply]
No, and actually `["src"]` does work, with all `#define`s correctly recognized, shown in the debug log, although I still receive errors that the header can not be found.  Now I'm so confused that I suspect that whether the error is given by fortls or others, e.g., the linter, which is `gfortran`.
